Kernel/Storage/NVMe/NVMeInterruptQueue.cpp

```diff
@@ -29,33 +29,51 @@ void NVMeInterruptQueue::submit_sqe(NVMeSubmission& sub)
     NVMeQueue::submit_sqe(sub);
 }
 
-void NVMeInterruptQueue::complete_current_request(u16 status)
+void NVMeInterruptQueue::complete_current_request(u16 cmdid, u16 status)
 {
-    VERIFY(m_request_lock.is_locked());
-
-    auto work_item_creation_result = g_io_work->try_queue([this, status]() {
+    auto work_item_creation_result = g_io_work->try_queue([this, cmdid, status]() {
         SpinlockLocker lock(m_request_lock);
-        auto current_request = m_current_request;
-        m_current_request.clear();
+        auto& request_pdu = m_requests.get(cmdid).release_value();
+        auto current_request = request_pdu.request;
+        request_pdu.used = false;
+        AsyncDeviceRequest::RequestResult req_result = AsyncDeviceRequest::Success;
+
+        // There can be submission without any request associated with it such as with
+        // admin queue commands during init. If there is no request, we are done
+        if (!current_request)
+            goto endio_handler;
+
         if (status) {
-            lock.unlock();
-            current_request->complete(AsyncBlockDeviceRequest::Failure);
-            return;
+            req_result = AsyncBlockDeviceRequest::Failure;
+            goto complete_request;
```

Comment:
Can we use some scope guards here instead? And some explicit scopes, if there's things we want to do *after* the spinlock has been released? Using goto and manually unlocking RAII objects is not a very C++ way to do this kind of operation.

Reply:
[https://imgs.xkcd.com/comics/goto.png](xkcd)

Reply:
> Can we use some scope guards here instead? And some explicit scopes, if there's things we want to do after the spinlock has been released? Using goto and manually unlocking RAII objects is not a very C++ way to do this kind of operation.
I felt the `goto` statements conveyed the intent of what I am trying to do. But I agree, I will take a look into ScopedGuard instead of goto. I think that should achieve what I want to do here without gotos